nexus/src/sagas.rs

```diff
@@ -453,30 +451,19 @@ async fn sic_create_custom_network_interfaces(
                     instance_id;
                     "primary_key" => interface_id.to_string(),
                 );
-
-                // Refetch the interface itself, to serialize it for the next
-                // saga node.
-                LookupPath::new(&opctx, &datastore)
-                    .instance_id(authz_instance.id())
-                    .network_interface_name(&db::model::Name(
-                        params.identity.name.clone(),
-                    ))
-                    .fetch()
-                    .await
-                    .map(|(.., db_interface)| db_interface)
+                Ok(())
             }
             Err(e) => Err(e.into_external()),
         }
         .map_err(ActionError::action_failed)?;
-        interfaces.push(NetworkInterface::from(interface))
     }
-    Ok(Some(interfaces))
+    Ok(())
 }
 
 /// Create the default network interface for an instance during the create saga
```

Comment:
That's correct, the "default" here is to create one interface, the primary, in the a VPC named `"default"` and VPC Subnet named `"default"`. This interface will have take the next available IP address in that VPC Subnet.

In this part of the code, we can refer to this as the primary interface, that seems fine. I used the term "default", which is absurdly overloaded, because this is the default outcome when one creates an instance. If you don't do anything else, you get a primary interface as described above. Some clarity around this name would be helpful. @rmustacc also brought up the point in last week's huddle that, by defaulting to a specific VPC and VPC Subnet, we probably want to give users the ability to control what those defaults are. For now, one would need to specify the VPC and VPC Subnet explicitly to get an interface created anywhere other than the one's named `"default"` in their Project.

Reply:
Yea I think it's very important to make sure we distinguish default from primary as they are really talking about two different things. A default is just something that affects creation time of the instance. A "default" interface is one that was selected semi-implicitly by the control plane. That is, it might offer up a drop down of subnets that the user could select from, with one of them filling the role of "default" as being the choice if you don't touch the dropdown. But even if you change the subnet selection there are still default values that can be chosen: e.g. the user could choose to let the IP be selected implicitly, rather than explicitly. These are all forms of "defaults".

On the other hand, the "primary" designator is a meaningful one not just at instance creation time, but also for the entire lifetime of the instance. The primary interface is the one that maps to DNS and external IPs. So it's important we make this distinction in the code and the UI, and I think we should call it "primary" and not default to avoid the confusion you already mentioned.

Reply:
I take your point. I've added [this issue](https://github.com/oxidecomputer/omicron/issues/960) to track cleaning this up, and possibly adding explicit tracking of the primary interface into Nexus and the database. That'd help when we start putting these IPs into DNS for example.

Reply:
Thanks for writing up that issue. I think we are covered here now.

Reply:
Thanks for writing that up. I still am uncertain if we should really have a default for provisioning versus a set of default things for VPC creation, but this is an important step in the split between the two. Thanks for capturing this @rzezeski and @bnaecker.

Reply:
So, a lot of the complexity here comes from the way I initially implemented the API for creating NICs for a new instance. There are currently [three variants](https://github.com/oxidecomputer/omicron/blob/main/nexus/src/external_api/params.rs#L96):

- `None` means create no interfaces at all
- `Default` means give me one primary interface in the default VPC and VPC Subnet, with an auto-assigned IP
- `Custom` takes zero or more instances of [this type](https://github.com/oxidecomputer/omicron/blob/c71e64d463a0b472db5982b591d83c3bac058fff/nexus/src/external_api/params.rs#L65), which are the parameters for any number of NICs.

The latter really covers the first two. I only added them to be more explicit about what is provided. I also thought there might be a variant to attach an _existing_ NIC, which doesn't make much sense, but pushed me towards this enum representation.

The point is, should we just remove that, and only accept a list of parameters for each NIC? If you want no NICs at all, pass zero items. If you want more than the one primary, just pass in the data for them. The defaults (determined in a TBD way) could be prepopulated in the console UI, which is closer in spirit to what Ryan mentioned earlier. We could also add a `primary` parameter, for users to specify which interface is used as the primary, or just use the first one as the primary.

Reply:
Let's move there as I'd like to focus on the different user cases and how we want to split up behavior between the API and clients around behavior here as there are arguments for the 'default' stuff at creation time, so I think it'll be good to take a step back from what we want the user experience to be here.